Fixes for synchronized client and transaction handling

thormick opened the pull request:
Fixes in the synchronized client and the parts of the transaction manager used by the synchronized client.

This has all the work I did in the sync client for [gapit.io](https://github.com/gapitio). Let me know if this should be split up, for instance with just the #538 fix. The other commits are for issues I discovered while digging to find and fix that issue.

Sorry for not getting around to this before, things have popped up so that even if this sat almost ready and just needed a once or twice over it still got punted a month, but at least now it's presentable enough for a PR.

No sweat about merging all of this, gapit.io is using a fork for pymodbus, has been doing so for a long time and can do so for a while longer.

thormick commented:
I see that the diagnostic client in acdd9e9e61aa4e122cb8864e4e9315211115c195 caused those issues, that can easily be backed out.

dhoomakethu commented on pymodbus/client/sync.py:
```diff
@@ -286,6 +294,35 @@ def _recv(self, size):
+        self.close()
+        readsize = (f"read of {size} bytes" if size
+                    else "unbounded read")
+        msg = (f'{self.__str__()}: Connection unexpectedly closed '
```
fstrings are supported only in python3, this would essentially the PR valid for pymodbus 3.x. If you intend to support to python 2.x or want this to go in to pymodbus 2.5.0 please update accordingly. Otherwise please raise this against branch 3.0.0

thormick replied:
Thank you, good catch, this was a copy from the custom branch that Gapit uses, which is only 3.0. I don't have any thought of whether this should go into the 2.x or 3.x versions. Might as well update it to work for 2.x.

dhoomakethu commented on pymodbus/client/sync.py:
```diff
@@ -256,9 +261,9 @@ def _recv(self, size):
+        # If size isn't specified read up to 4096 bytes at a time.
         if size is None:
-            recv_size = 1
+            recv_size = 4096
```
This value is debatable. This would essentially block the `recv` till all `4096` bytes are received or timeout. Adding further to delay in reads.

thormick replied:
That's not how `recv` works. From the documentation:
```
$ pydoc socket.socket.recv | more
Help on method_descriptor in socket.socket:

socket.socket.recv = recv(...)
    recv(buffersize[, flags]) -> data
    
    Receive up to buffersize bytes from the socket.  For the optional flags
    argument, see the Unix manual.  When no data is available, block until
    at least one byte is available or until the remote end is closed.  When
    the remote end is closed and all data is read, return the empty string.
```

So it never blocks when a single byte is available, regardless of what `buffersize` is provided. I.e. it's a maximum limit, the minimum limit is always 1 regardless.

tracernz replied:
@thormick That matches the underlying recv man page too:

https://man7.org/linux/man-pages/man2/recv.2.html
> If no messages are available at the socket, the receive calls wait for a message to arrive, unless the socket is nonblocking (see fcntl(2)), in which case the value -1 is returned and the external variable errno is set to EAGAIN or EWOULDBLOCK.  The receive calls normally return any data available, up to the requested amount, rather than waiting for receipt of the full amount requested.

dhoomakethu replied:
ah, I got this confused with `Serial`.Apologies for the mistake. Looks good.

thormick replied:
Actually, it's good that you picked on it, it made me gave it a quick look over and I realize now that `ModbusTlsClient` does the same, and now those two will be inconsistent. I don't have a setup ready for testing modbus TLS, so I'm not sure about touching that part of the code. Not necessarily a problem, though?

thormick commented:
I've got the fstring thing sorted out, added and changed the unit tests a bit, and done a few other fixes too. But I'm not sure about intricacies regarding GitHub; is it correct to force push to the branch this PR is based off? Or should this PR be closed and another opened? The revised version resides in the branch at https://github.com/gapitio/pymodbus/tree/fix-sync-client-trans-2 for now.

dhoomakethu commented:
@thormick you can push the changes to the same PR, I will cherry-pick the changes to `2.5.0` and from there it gets merged to `dev` and subsequently to `master` once `2.5.0` is officially released. Currently I have the  release candidate `2.5.0rc2` 
 on pypi. May be this PR will go with `2.5.0rc3`.

thormick commented:
@dhoomakethu So done, I've rammed it in now. The major difference from the previous iteration is all the changes to the test_client_sync.py. Most of it is just to increase coverage enough that the CI doesn't complain, but I've also included the admittedly highly brazen change of patching out `time`. This will hopefully make the test more deterministic, and in my tests more than halves the time it takes to run the entire unittest suite. I can back that out again if it's too radical.

thormick commented:
Made some small additions to the unit tests. Commit-to-commit diff relative to my Jan 8 push [here](https://github.com/gapitio/pymodbus/compare/gapitio:946297439374d8f11b1805174d534394e4c6000f..gapitio:24c0ffc975c40a9dcab9620847a195f5565eaad3).

dhoomakethu commented:
@thormick I have merged the changes in this PR in #605 . Will be closing this PR.

thormick commented:
@dhoomakethu Thank you very much, glad you could make use of it all.
